Source/WebCore/animation/ViewTimeline.cpp

```diff
@@ -138,4 +142,75 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ViewTimeline::documentWil
     return AnimationTimeline::ShouldUpdateAnimationsAndSendEvents::No;
 }
 
+Element* ViewTimeline::source() const
+{
+    if (auto* sourceRender = sourceRenderer())
+        return sourceRender->element();
+    return nullptr;
+}
+
+RenderBox* ViewTimeline::sourceRenderer() const
+{
+    if (!m_subject)
+        return nullptr;
+
+    Ref document = m_subject->document();
+
+    CheckedPtr subjectRenderer = m_subject->renderer();
+    if (!subjectRenderer)
+        return { };
+
+    // https://drafts.csswg.org/scroll-animations-1/#dom-scrolltimeline-source
+    // Determine source renderer by looking for the nearest ancestor that establishes a scroll container
+    return subjectRenderer->enclosingScrollableContainerForSnapping();
+}
+
+ViewTimeline::ViewTimelineData ViewTimeline::computeViewTimelineData() const
+{
+    if (!m_subject)
+        return { };
+
+    Ref document = m_subject->document();
+
+    CheckedPtr subjectRenderBox = dynamicDowncast<RenderBox>(m_subject->renderer());
+    if (!subjectRenderBox)
+        return { };
+
+    CheckedPtr<RenderBox> sourceRenderBox = sourceRenderer();
+
+    if (!sourceRenderBox)
+        return { };
+
+    // https://drafts.csswg.org/scroll-animations-1/#view-timeline-progress
+    //
+    // Progress (the current time) in a view progress timeline is calculated as: distance ÷ range where:
+    //
+    // - distance is the current scroll offset minus the scroll offset corresponding to the start of the
+    //   cover range
+    // - range is the scroll offset corresponding to the start of the cover range minus the scroll offset
+    //   corresponding to the end of the cover range
+    // TODO: take into account view-timeline-inset: https://drafts.csswg.org/scroll-animations-1/#propdef-view-timeline-inset
+    // TODO: take into account animation-range: https://drafts.csswg.org/scroll-animations-1/#animation-range
+
+    float currentScrollOffset = 0;
+    float scrollContainerSize;
+    if (sourceRenderBox->element() == document->documentElement()) {
+        auto& frameView = sourceRenderBox->view().frameView();
```

Comment:
FrameView is a ScrollableArea so we should be able to write this code in terms of ScrollableArea, instead of branching.

Reply:
Fixed.